apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs

```diff
@@ -0,0 +1,393 @@
+//! Code for handling [`ProtocolTxType::EthereumEvents`] transactions.
+mod events;
+mod read;
+mod update;
+mod utils;
+
+use std::collections::{BTreeSet, HashMap, HashSet};
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use eyre::{eyre, Result};
+use namada::ledger::eth_bridge::storage::eth_msgs::{self, Keys};
+use namada::ledger::pos::types::{VotingPower, WeightedValidator};
+use namada::types::address::Address;
+use namada::types::ethereum_events::EthereumEvent;
+use namada::types::storage::{self, Epoch};
+use namada::types::transaction::TxResult;
+use namada::types::vote_extensions::ethereum_events::MultiSignedEthEvent;
+use num_rational::Ratio;
+
+use super::store::{Store, StoreExt};
+
+/// Represents an event stored under `eth_msgs`
+#[derive(
+    Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct EthMsg {
+    /// The event being stored
+    pub body: EthereumEvent,
+    /// The total voting power that's voted for this event across all epochs
+    pub voting_power: (u64, u64),
+    /// The addresses of validators that voted for this event, in sorted order.
+    pub seen_by: Vec<Address>,
```

Comment:
you can pick a `HashSet` or `BTreeMap` here, even if we need deterministic data in storage. borsh should guarantee we get the same encoding every time.

Reply:
I agree, we should use a `BTreeSet`. I would rather make this change in a new PR though (and update the specs at the same time), I can make an issue for it

Reply:
https://github.com/anoma/namada/issues/434